tests/experimental/nn/test_auto_shard.py

```diff
@@ -107,3 +105,39 @@ def test_single_run():
             input = model(input)
 
     assert input.size() == torch.Size([35, 20, 28783])
+
+
+class Branch(torch.nn.Module):
+    def __init__(self, features: int = 10):
+        super().__init__()
+        self.left = nn.Linear(in_features=features, out_features=features)
+        self.right = nn.Linear(in_features=features, out_features=features)
+
+    def forward(self, x):
+        if x.sum() > 1000:
+            return self.left(x)
+        else:
+            return self.right(x)
+
+
+class BranchedNetwork(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.net = torch.nn.ModuleList([Branch() for _ in range(10)])
+
+    def forward(self, x):
+        for module in self.net:
+            x = module(x)
+        return x
+
+
+def test_dynaimc_conditionals_auto_wrapped():
+    if torch_version() < (1, 8, 0):
+        pytest.skip("requires torch version >= 1.8.0")
+    from fairscale.experimental.nn.auto_shard import shard_model
+
+    model = BranchedNetwork()
+    sharded_model = shard_model(model, 3)
```

Comment:
can we test a quick FW pass to make sure we see the same result when compared to unsharded?

Reply:
Done.